File: src/main/java/com/qwazr/jdbc/cache/CachedStatement.java
/*
 * Copyright 2016-2017 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.qwazr.jdbc.cache;

import javax.xml.bind.DatatypeConverter;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.SQLFeatureNotSupportedException;
import java.sql.SQLWarning;
import java.sql.Statement;

class CachedStatement<T extends Statement> implements Statement {

    private final CachedConnection connection;
    final ResultSetCache resultSetCache;
    final T backendStatement;

    private final int resultSetConcurrency;
    private final int resultSetType;
    private final int resultSetHoldability;

    private volatile int maxFieldSize;
    private volatile int maxRows;
    private volatile int queryTimeOut;
    private volatile int fetchDirection;
    private volatile int fetchSize;
    private volatile boolean closed;
    private volatile boolean poolable;
    private volatile boolean closeOnCompletion;

    volatile String executedSql;
    volatile String generatedKey;

    CachedStatement(final CachedConnection connection, final ResultSetCache resultSetCache,
            final T backendStatement, final int resultSetConcurrency, final int resultSetType,
            final int resultSetHoldability) {
        this.connection = connection;
        this.resultSetCache = resultSetCache;
        this.backendStatement = backendStatement;
        this.resultSetConcurrency = resultSetConcurrency;
        this.resultSetType = resultSetType;
        this.resultSetHoldability = resultSetHoldability;
        this.maxFieldSize = 0;
        this.maxRows = 0;
        this.queryTimeOut = 0;
        this.fetchDirection = 0;
        this.fetchSize = 0;
        this.closed = false;
        this.poolable = false;
        this.closeOnCompletion = false;
        this.executedSql = null;
    }

    CachedStatement(final CachedConnection connection, final ResultSetCache resultSetCache,
            final T backendStatement) {
        this(connection, resultSetCache, backendStatement, 0, 0, 0);
    }

    final T checkBackendStatement() throws SQLException {
        return checkBackendStatement(null);
    }

    final T checkBackendStatement(final String error) throws SQLException {
        if (backendStatement != null)
            return backendStatement;
        else
            throw error == null ? new SQLFeatureNotSupportedException() : new SQLException(error);
    }

    static String generateCacheKey(final String src) throws SQLException {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            return DatatypeConverter.printHexBinary(md.digest(src.getBytes()));
        } catch (NoSuchAlgorithmException e) {
            throw new SQLException("MD5 is not available");
        }
    }

    protected void generateKey() throws SQLException {
        generatedKey = generateCacheKey(executedSql);
    }

    final String getOrGenerateKey() throws SQLException {
        if (generatedKey == null)
            generateKey();
        return generatedKey;
    }

    @Override
    public ResultSet executeQuery(String sql) throws SQLException {
        this.executedSql = sql;
        generateKey();
        return resultSetCache
                .get(this, generatedKey, backendStatement == null ? null : () -> backendStatement.executeQuery(sql));
    }

    @Override
    public int executeUpdate(String sql) throws SQLException {
        this.executedSql = sql;
        return checkBackendStatement().executeUpdate(sql);
    }

    @Override
    public void close() throws SQLException {
        if (backendStatement != null)
            backendStatement.close();
        closed = true;
    }

    @Override
    public int getMaxFieldSize() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getMaxFieldSize();
        else
            return maxFieldSize;
    }

    @Override
    public void setMaxFieldSize(int max) throws SQLException {
        if (backendStatement != null)
            backendStatement.setMaxFieldSize(max);
        this.maxFieldSize = max;
    }

    @Override
    public int getMaxRows() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getMaxRows();
        else
            return maxRows;
    }

    @Override
    public void setMaxRows(int max) throws SQLException {
        if (backendStatement != null)
            backendStatement.setMaxRows(max);
        this.maxRows = max;
    }

    @Override
    public void setEscapeProcessing(boolean enable) throws SQLException {
        if (backendStatement != null)
            backendStatement.setEscapeProcessing(enable);
    }

    @Override
    public int getQueryTimeout() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getQueryTimeout();
        else
            return queryTimeOut;
    }

    @Override
    public void setQueryTimeout(int seconds) throws SQLException {
        if (backendStatement != null)
            backendStatement.setQueryTimeout(seconds);
        this.queryTimeOut = seconds;
    }

    @Override
    public void cancel() throws SQLException {
        if (backendStatement != null)
            backendStatement.cancel();
    }

    @Override
    public SQLWarning getWarnings() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getWarnings();
        else
            return null;
    }

    @Override
    public void clearWarnings() throws SQLException {
        if (backendStatement != null)
            backendStatement.clearWarnings();
    }

    @Override
    public void setCursorName(String name) throws SQLException {
        if (backendStatement != null)
            backendStatement.setCursorName(name);
    }

    @Override
    public boolean execute(String sql) throws SQLException {
        this.executedSql = sql;
        generateKey();
        return resultSetCache.checkIfExists(generatedKey) || checkBackendStatement("No cache entry").execute(sql);
    }

    @Override
    final public ResultSet getResultSet() throws SQLException {
        generateKey();
        return resultSetCache.get(this, generatedKey, backendStatement == null ? null : backendStatement::getResultSet);
    }

    @Override
    public int getUpdateCount() throws SQLException {
        return checkBackendStatement().getUpdateCount();
    }

    @Override
    public boolean getMoreResults() throws SQLException {
        return checkBackendStatement().getMoreResults();
    }

    @Override
    public void setFetchDirection(int direction) throws SQLException {
        if (backendStatement != null)
            backendStatement.setFetchDirection(direction);
        this.fetchDirection = direction;
    }

    @Override
    public int getFetchDirection() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getFetchDirection();
        else
            return fetchDirection;
    }

    @Override
    public void setFetchSize(int rows) throws SQLException {
        if (backendStatement != null)
            backendStatement.setFetchSize(rows);
        this.fetchSize = rows;
    }

    @Override
    public int getFetchSize() throws SQLException {
        if (backendStatement != null)
            return backendStatement.getFetchSize();
        else
            return fetchSize;
    }

    @Override
    public int getResultSetConcurrency() throws SQLException {
        return resultSetConcurrency;
    }

    @Override
    public int getResultSetType() throws SQLException {
        return resultSetType;
    }

    @Override
    public void addBatch(String sql) throws SQLException {
        checkBackendStatement().addBatch(sql);
    }

    @Override
    public void clearBatch() throws SQLException {
        checkBackendStatement().clearBatch();
    }

    @Override
    public int[] executeBatch() throws SQLException {
        return checkBackendStatement().executeBatch();
    }

    @Override
    public Connection getConnection() throws SQLException {
        return connection;
    }

    @Override
    public boolean getMoreResults(int current) throws SQLException {
        return checkBackendStatement().getMoreResults(current);
    }

    @Override
    public ResultSet getGeneratedKeys() throws SQLException {
        return checkBackendStatement().getGeneratedKeys();
    }

    @Override
    public int executeUpdate(String sql, int autoGeneratedKeys) throws SQLException {
        executedSql = sql;
        return checkBackendStatement().executeUpdate(sql, autoGeneratedKeys);
    }

    @Override
    public int executeUpdate(String sql, int[] columnIndexes) throws SQLException {
        executedSql = sql;
        return checkBackendStatement().executeUpdate(sql, columnIndexes);
    }

    @Override
    public int executeUpdate(String sql, String[] columnNames) throws SQLException {
        executedSql = sql;
        return checkBackendStatement().executeUpdate(sql, columnNames);
    }

    @Override
    public boolean execute(String sql, int autoGeneratedKeys) throws SQLException {
        executedSql = sql;
        generateKey();
        return resultSetCache.checkIfExists(generatedKey) || checkBackendStatement("No cache entry")
                .execute(sql, autoGeneratedKeys);
    }

    @Override
    public boolean execute(String sql, int[] columnIndexes) throws SQLException {
        executedSql = sql;
        generateKey();
        return resultSetCache.checkIfExists(generatedKey) || checkBackendStatement("No cache entry")
                .execute(sql, columnIndexes);
    }

    @Override
    public boolean execute(String sql, String[] columnNames) throws SQLException {
        executedSql = sql;
        generateKey();
        return resultSetCache.checkIfExists(generatedKey) || checkBackendStatement("No cache entry")
                .execute(sql, columnNames);
    }

    @Override
    public int getResultSetHoldability() throws SQLException {
        return resultSetHoldability;
    }

    @Override
    public boolean isClosed() throws SQLException {
        if (backendStatement != null)
            return backendStatement.isClosed();
        else
            return closed;
    }

    @Override
    public void setPoolable(boolean poolable) throws SQLException {
        if (backendStatement != null)
            backendStatement.setPoolable(poolable);
        this.poolable = poolable;
    }

    @Override
    public boolean isPoolable() throws SQLException {
        if (backendStatement != null)
            return backendStatement.isPoolable();
        else
            return poolable;
    }

    @Override
    public void closeOnCompletion() throws SQLException {
        if (backendStatement != null)
            backendStatement.closeOnCompletion();
        this.closeOnCompletion = true;
    }

    @Override
    public boolean isCloseOnCompletion() throws SQLException {
        if (backendStatement != null)
            return backendStatement.isCloseOnCompletion();
        else
            return closeOnCompletion;
    }

    @Override
    public <V> V unwrap(Class<V> iface) throws SQLException {
        return checkBackendStatement().unwrap(iface);
    }

    @Override
    public boolean isWrapperFor(Class<?> iface) throws SQLException {
        return checkBackendStatement().isWrapperFor(iface);
    }
}

File: src/main/kotlin/io/github/jhstatewide/jdbc/cache/CachedStatement.kt
package io.github.jhstatewide.jdbc.cache/*
 * Copyright 2016-2017 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.sql.*

internal const val JDBC_CACHE_EXTENSION = ".jdbc_cache"

open class CachedStatement<T : Statement?> @JvmOverloads constructor(
    private val connection: CachedConnection?, val resultSetCache: ResultSetCache?,
    backendStatement: T, resultSetConcurrency: Int = 0, resultSetType: Int = 0,
    resultSetHoldability: Int = 0
) : Statement {
    val backendStatement: T?
    private val resultSetConcurrency: Int
    private val resultSetType: Int
    private val resultSetHoldability: Int

    @Volatile
    private var maxFieldSize: Int

    @Volatile
    private var maxRows: Int

    @Volatile
    private var queryTimeOut: Int

    @Volatile
    private var fetchDirection: Int

    @Volatile
    private var fetchSize: Int

    @Volatile
    private var closed: Boolean

    @Volatile
    private var poolable: Boolean

    @Volatile
    private var closeOnCompletion: Boolean

    @Volatile
    var executedSql: String?

    @Volatile
    var generatedKey: String? = null

    init {
        this.backendStatement = backendStatement
        this.resultSetConcurrency = resultSetConcurrency
        this.resultSetType = resultSetType
        this.resultSetHoldability = resultSetHoldability
        maxFieldSize = 0
        maxRows = 0
        queryTimeOut = 0
        fetchDirection = 0
        fetchSize = 0
        closed = false
        poolable = false
        closeOnCompletion = false
        executedSql = null
    }

    @JvmOverloads
    @Throws(SQLException::class)
    fun checkBackendStatement(error: String? = null): T {
        return backendStatement ?: throw error?.let { SQLException(it) } ?: SQLFeatureNotSupportedException()
    }

    @Throws(SQLException::class)
    protected open fun generateKey() {
        generatedKey = generateCacheKey(executedSql)
    }

    @get:Throws(SQLException::class)
    val orGenerateKey: String?
        get() {
            if (generatedKey == null) generateKey()
            return generatedKey
        }

    @Throws(SQLException::class)
    override fun executeQuery(sql: String): ResultSet? {
        executedSql = sql
        generateKey()
        return resultSetCache
            ?.get<Statement>(
                this,
                generatedKey,
                if (backendStatement == null) {
                    null
                } else {
                    object : ResultSetCache.Provider {
                        override fun provide(): ResultSet? {
                            return backendStatement.executeQuery(sql)
                        }
                    }
                }
            )
    }

    @Throws(SQLException::class)
    override fun executeUpdate(sql: String): Int {
        executedSql = sql
        return checkBackendStatement()!!.executeUpdate(sql)
    }

    @Throws(SQLException::class)
    override fun close() {
        backendStatement?.close()
        closed = true
    }

    @Throws(SQLException::class)
    override fun getMaxFieldSize(): Int {
        return backendStatement?.maxFieldSize ?: maxFieldSize
    }

    @Throws(SQLException::class)
    override fun setMaxFieldSize(max: Int) {
        if (backendStatement != null) backendStatement.maxFieldSize = max
        maxFieldSize = max
    }

    @Throws(SQLException::class)
    override fun getMaxRows(): Int {
        return backendStatement?.maxRows ?: maxRows
    }

    @Throws(SQLException::class)
    override fun setMaxRows(max: Int) {
        if (backendStatement != null) backendStatement.maxRows = max
        maxRows = max
    }

    @Throws(SQLException::class)
    override fun setEscapeProcessing(enable: Boolean) {
        backendStatement?.setEscapeProcessing(enable)
    }

    @Throws(SQLException::class)
    override fun getQueryTimeout(): Int {
        return backendStatement?.queryTimeout ?: queryTimeOut
    }

    @Throws(SQLException::class)
    override fun setQueryTimeout(seconds: Int) {
        if (backendStatement != null) backendStatement.queryTimeout = seconds
        queryTimeOut = seconds
    }

    @Throws(SQLException::class)
    override fun cancel() {
        backendStatement?.cancel()
    }

    @Throws(SQLException::class)
    override fun getWarnings(): SQLWarning? {
        return backendStatement?.warnings
    }

    @Throws(SQLException::class)
    override fun clearWarnings() {
        backendStatement?.clearWarnings()
    }

    @Throws(SQLException::class)
    override fun setCursorName(name: String) {
        backendStatement?.setCursorName(name)
    }

    @Throws(SQLException::class)
    override fun execute(sql: String): Boolean {
        executedSql = sql
        generateKey()
        return resultSetCache?.checkIfExists(generatedKey) == true || checkBackendStatement("No cache entry")!!.execute(sql)
    }

    @Throws(SQLException::class)
    override fun getResultSet(): ResultSet? {
        generateKey()
        return resultSetCache?.get<Statement>(
            this,
            generatedKey,
            if (backendStatement == null) {
                null
            } else {
                object : ResultSetCache.Provider {
                    override fun provide(): ResultSet? {
                        return backendStatement.resultSet
                    }
                }
            }
        )
    }

    @Throws(SQLException::class)
    override fun getUpdateCount(): Int {
        return checkBackendStatement()!!.updateCount
    }

    @Throws(SQLException::class)
    override fun getMoreResults(): Boolean {
        return checkBackendStatement()!!.moreResults
    }

    @Throws(SQLException::class)
    override fun setFetchDirection(direction: Int) {
        if (backendStatement != null) backendStatement.fetchDirection = direction
        fetchDirection = direction
    }

    @Throws(SQLException::class)
    override fun getFetchDirection(): Int {
        return backendStatement?.fetchDirection ?: fetchDirection
    }

    @Throws(SQLException::class)
    override fun setFetchSize(rows: Int) {
        if (backendStatement != null) backendStatement.fetchSize = rows
        fetchSize = rows
    }

    @Throws(SQLException::class)
    override fun getFetchSize(): Int {
        return backendStatement?.fetchSize ?: fetchSize
    }

    @Throws(SQLException::class)
    override fun getResultSetConcurrency(): Int {
        return resultSetConcurrency
    }

    @Throws(SQLException::class)
    override fun getResultSetType(): Int {
        return resultSetType
    }

    @Throws(SQLException::class)
    override fun addBatch(sql: String) {
        checkBackendStatement()!!.addBatch(sql)
    }

    @Throws(SQLException::class)
    override fun clearBatch() {
        checkBackendStatement()!!.clearBatch()
    }

    @Throws(SQLException::class)
    override fun executeBatch(): IntArray {
        return checkBackendStatement()!!.executeBatch()
    }

    @Throws(SQLException::class)
    override fun getConnection(): CachedConnection? {
        return connection
    }

    @Throws(SQLException::class)
    override fun getMoreResults(current: Int): Boolean {
        return checkBackendStatement()!!.getMoreResults(current)
    }

    @Throws(SQLException::class)
    override fun getGeneratedKeys(): ResultSet {
        return checkBackendStatement()!!.generatedKeys
    }

    @Throws(SQLException::class)
    override fun executeUpdate(sql: String, autoGeneratedKeys: Int): Int {
        executedSql = sql
        return checkBackendStatement()!!.executeUpdate(sql, autoGeneratedKeys)
    }

    @Throws(SQLException::class)
    override fun executeUpdate(sql: String, columnIndexes: IntArray): Int {
        executedSql = sql
        return checkBackendStatement()!!.executeUpdate(sql, columnIndexes)
    }

    @Throws(SQLException::class)
    override fun executeUpdate(sql: String, columnNames: Array<String>): Int {
        executedSql = sql
        return checkBackendStatement()!!.executeUpdate(sql, columnNames)
    }

    @Throws(SQLException::class)
    override fun execute(sql: String, autoGeneratedKeys: Int): Boolean {
        executedSql = sql
        generateKey()
        return resultSetCache?.checkIfExists(generatedKey) == true || checkBackendStatement("No cache entry")
            ?.execute(sql, autoGeneratedKeys) ?: false
    }

    @Throws(SQLException::class)
    override fun execute(sql: String, columnIndexes: IntArray): Boolean {
        executedSql = sql
        generateKey()
        return resultSetCache?.checkIfExists(generatedKey) == true || checkBackendStatement("No cache entry")
            ?.execute(sql, columnIndexes) ?: false
    }

    @Throws(SQLException::class)
    override fun execute(sql: String, columnNames: Array<String>): Boolean {
        executedSql = sql
        generateKey()
        return resultSetCache?.checkIfExists(generatedKey) == true || checkBackendStatement("No cache entry")
            ?.execute(sql, columnNames) ?: false
    }

    @Throws(SQLException::class)
    override fun getResultSetHoldability(): Int {
        return resultSetHoldability
    }

    @Throws(SQLException::class)
    override fun isClosed(): Boolean {
        return backendStatement?.isClosed ?: closed
    }

    @Throws(SQLException::class)
    override fun setPoolable(poolable: Boolean) {
        if (backendStatement != null) backendStatement.isPoolable = poolable
        this.poolable = poolable
    }

    @Throws(SQLException::class)
    override fun isPoolable(): Boolean {
        return backendStatement?.isPoolable ?: poolable
    }

    @Throws(SQLException::class)
    override fun closeOnCompletion() {
        backendStatement?.closeOnCompletion()
        closeOnCompletion = true
    }

    @Throws(SQLException::class)
    override fun isCloseOnCompletion(): Boolean {
        return backendStatement?.isCloseOnCompletion ?: closeOnCompletion
    }

    @Throws(SQLException::class)
    override fun <V> unwrap(iface: Class<V>): V {
        return checkBackendStatement()!!.unwrap(iface)
    }

    @Throws(SQLException::class)
    override fun isWrapperFor(iface: Class<*>?): Boolean {
        return checkBackendStatement()!!.isWrapperFor(iface)
    }

    companion object {
        @Throws(SQLException::class)
        fun generateCacheKey(src: String?): String {
            return try {
                val md = MessageDigest.getInstance("MD5")
                DatatypeConverter.printHexBinary(md.digest(src!!.toByteArray())) + JDBC_CACHE_EXTENSION
            } catch (e: NoSuchAlgorithmException) {
                throw SQLException("MD5 is not available")
            }
        }
    }
}